### PR TITLE
Enhancements to FloatingViewportComponent

### PR DESCRIPTION
**Pull request by a-gagnon**
### FloatingViewportComponent
Add viewportRef in the props
Change the initialViewState to also accept a function that returns a viewState rather than the viewState itself.
Removed null checks for the `initialViewState` as it can't possibly be null.

### CODE REVIEW

**Comment by a-gagnon** [comment]
@NancyMcCallB I tried to use the `FloatingViewportContent` as a drop-in replacement for the `ViewportComponent` but our code requires that we grab the viewport ref. I wasn't able to do that with the `props` that are available.

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx [hunk 1]
```diff
@@ -39,11 +41,17 @@ export interface FloatingViewportContentProps {
...
-  const { contentId, initialViewState } = props;
+  const { contentId, initialViewState, viewportRef } = props;
   const [viewport, setViewport] = React.useState<ScreenViewport | undefined>();
   const contentControl = React.useRef<FloatingViewportContentControl | undefined>();
 
-  const onViewportRef = React.useCallback((v: ScreenViewport) => setViewport(v), []);
+  const viewState = React.useMemo(() => typeof initialViewState === "function" ? initialViewState() : initialViewState, [initialViewState]);
+  const onViewportRef = React.useCallback((v: ScreenViewport) => {
+    setViewport(v);
+    if (viewportRef)
```
Comment (raplemie) [inline ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:51]: Please add a test for this new branch.

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx [hunk 2]
```diff
@@ -27,8 +27,10 @@ export interface FloatingViewportContentProps {
   onContextMenu?: (e: React.MouseEvent) => boolean;
   /** viewport/content control uniqueId */
   contentId: string;
-  /** The initial view state used to create the viewport */
-  initialViewState: ViewState;
+  /** The initial view state used to create the viewport, or a function that returns it */
```
Comment (raplemie) [inline ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:30]:
```suggestion
  /** The initial view state used to create the viewport, or a function that returns it (will refresh when the function changes) */
```

**Review by raplemie — CHANGES_REQUESTED** [review]
I was wondering if we shouldnt just pass in the viewState through for FloatingViewport, but we sadly need the viewstate resolved for the processViewSelectorChange...

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx (hunk 1 above)
Comment (raplemie) [reply ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:51]: I just saw that the whole component is marked as istanbul ignore, so we'll need to jump in and write test, don't start this from scratch.

**Comment by a-gagnon** [comment]
> I was wondering if we shouldnt just pass in the viewState through for FloatingViewport, but we sadly need the viewstate resolved for the processViewSelectorChange...

Right. My main concern is to make the props object look similar to the ViewportComponentProps.

**Review by raplemie — APPROVED** [review]
LGTM

**Review by NancyMcCallB — APPROVED** [review]

File: common/api/appui-react.api.md [hunk 3]
```diff
@@ -2181,8 +2181,9 @@ export class FloatingViewportContentControl extends ViewportContentControl {
 // @beta (undocumented)
 export interface FloatingViewportContentProps {
     contentId: string;
-    initialViewState: ViewState;
+    initialViewState: ViewStateProp;
     onContextMenu?: (e: React_2.MouseEvent) => boolean;
+    viewportRef?: (v: ScreenViewport) => void;
```
Comment (GerardasB) [inline common/api/appui-react.api.md:2186]: Could we use an actual `React.Ref<ScreenViewport>` type here? Even though the `ViewportProps["viewportRef"]` is currently defined as a callback...

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx [hunk 4]
```diff
@@ -39,11 +41,22 @@ export interface FloatingViewportContentProps {
...
+  const onViewportRef = React.useCallback((v: ScreenViewport) => {
+    setViewport(v);
+    if (!viewportRef)
+      return;
+
+    if (typeof viewportRef === "function")
+      viewportRef(v);
+    else
+      viewportRef.current = v;
+
```
Comment (raplemie) [inline ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:58]: Use braces even for single line content.
```suggestion
    if (!viewportRef) {
      return;
    }
    if (typeof viewportRef === "function") {
      viewportRef(v);
    } else {
      viewportRef.current = v;
    }
```

**Review by raplemie — DISMISSED** [review]

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx [hunk 5]
```diff
@@ -27,8 +27,10 @@ export interface FloatingViewportContentProps {
   onContextMenu?: (e: React.MouseEvent) => boolean;
   /** viewport/content control uniqueId */
   contentId: string;
-  /** The initial view state used to create the viewport */
-  initialViewState: ViewState;
+  /** The initial view state used to create the viewport, or a function that returns it (will refresh when the function changes) */
+  initialViewState: ViewStateProp;
+  /** Function to get a reference to the ScreenViewport */
+  viewportRef?: React.ForwardedRef<ScreenViewport>;
```
Comment (GerardasB) [inline ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:33]: I think we should keep this as `React.Ref<ScreenViewport>`, to merge the `viewportRef` w/ `onViewportRef` you could use `useRefs` or `mergeRefs` from `core-react`..

Comment (raplemie) [reply ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:33]: Ref are immutable, so we cant assing to it, so I suggested the ForwardedRef...

**Comment by a-gagnon** [comment]
Reviving this PR that has more or less been put on a shelf for the past weeks.

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx (hunk 5 above)
Comment (raplemie) [reply ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:33]: After looking around a bit, I think this type is indeed not meant to be used like that, so my suggestion would be wrong, That being said, React "ref" are not meant to be props and are handled differently, so I guess we might as well just use Gerardas suggestion (or revert back to what it was before, as these are not real refs, the goal being to have the same behavior as ViewportComponent ?) @GerardasB  ?

Comment (GerardasB) [reply ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:33]: Yes, I guess we can match what's currently in the `ViewportComponent`. We can refactor both components later.

Comment (GerardasB) [inline ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:33]:
```suggestion
  viewportRef?: React.Ref<ScreenViewport>;
```

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx [hunk 6]
```diff
@@ -39,11 +41,22 @@ export interface FloatingViewportContentProps {
...
+    setViewport(v);
+    if (!viewportRef) {
+      return;
+    }
+    if (typeof viewportRef === "function") {
+      viewportRef(v);
+    } else {
+      viewportRef.current = v;
+    }
+  }, [viewportRef]);
```
Comment (GerardasB) [inline ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:59]: Would this work?
```suggestion
  const ref = React.useCallback((v: ScreenViewport) => {
    setViewport(v);
  }, []);
  const onViewportRef = useRefs(ref, ...(viewportRef ? [viewportRef] : []));
```

**Review by GerardasB — CHANGES_REQUESTED** [review]

File: ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx (hunk 6 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/content/FloatingViewportContent.tsx:59]: We should probably allow `undefined` in `useRefs` to simplify cases like these...

**Pull request merged**